### ISSUE TITLE
MLX Follow-Up: Mixtral Router Adapter

### ISSUE THREAD

**Issue by egesabanci** [egesabanci/reap-mlx#10]
## Objective

Add Mixtral routing support to the MLX backend after the Qwen3-focused pruning path is working.

## Background

Mixtral routing differs from Qwen3 routing. Qwen3 applies full softmax over all experts and selects top-k probabilities. Mixtral selects top-k over raw logits and then applies softmax only over selected logits. Using a generic Qwen-style router for Mixtral would produce incorrect selected scores and possibly different pruning saliency.

## Scope

Add Mixtral router and adapter/config coverage. This is a follow-up issue and should not block the first Qwen3-compatible MLX path.

Target files may include:

- `src/reap/backends/mlx/router.py`
- `src/reap/backends/mlx/model_adapters.py`
- `src/reap/backends/mlx/prune.py`
- tests such as `tests/test_mlx_mixtral_router.py`

## Dependencies

This issue should start after the Qwen3 MLX path has router, metrics, model adapter, observer, and pruning basics in place.

## Implementation Steps

1. Inspect the installed/current MLX-LM Mixtral implementation for exact routing behavior.
2. Add `MixtralRouter` that follows Mixtral semantics:
   - call `moe.gate(x)` or the architecture-specific gate
   - select indices using `mx.argpartition(-logits, kth=top_k - 1, axis=-1)[..., :top_k]`
   - gather selected logits
   - compute selected scores with `mx.softmax(selected_logits, axis=-1, precise=True)`
3. Add tests comparing against a hand-written Mixtral reference implementation.
4. Add Mixtral adapter/config mapping:
   - expert count key: `num_local_experts`
   - top-k key: `num_experts_per_tok`
   - MoE attribute: typically `block_sparse_moe` or adapter-discovered equivalent
5. Extend pruning config update logic for Mixtral-specific fields.
6. Verify stacked expert tensor slicing paths for Mixtral MLX-LM modules.
7. Add a small synthetic Mixtral-style model test if possible without downloading real weights.
8. Add a manual/slow real-model smoke command if an MLX-LM Mixtral model is available.

## Acceptance Criteria

- [ ] `MixtralRouter` is implemented and importable.
- [ ] Mixtral router tests validate raw-logit top-k then selected-softmax behavior.
- [ ] Mixtral routing tests do not use Qwen full-softmax semantics.
- [ ] Mixtral config/expert-count/top-k fields are updated correctly after pruning.
- [ ] Tiny/synthetic Mixtral-style pruning shape tests pass.
- [ ] Real-model validation is documented as manual/slow, not default CI.
- [ ] Importing Mixtral support does not import `torch` or `vllm`.

## Test Plan

Run:

```bash
pytest tests/test_mlx_no_torch_import.py tests/test_mlx_mixtral_router.py tests/test_mlx_prune.py -v
```

Manual/slow command shape if a suitable model is available:

```bash
python -m reap.backends.mlx.entrypoint \
  --model-name <mlx-mixtral-model> \
  --dataset-name theblackcat102/evol-codealpaca-v1 \
  --prune-method reap \
  --compression-ratio 0.25 \
  --num-calibration-sequences 8 \
  --output-dir artifacts/mlx/mixtral-smoke
```

## Non-Goals

- Do not add DeepSeek, GLM, or ERNIE routing in this issue.
- Do not implement merge metrics.
- Do not run full benchmark evaluation.
- Do not refactor the architecture adapter layer beyond what Mixtral needs.

## References

- `implementation-plan-draft.md`, Phase 4
- `docs/reap-mlx-implementation-plan-gpt-v0.2.md`, Mixtral Router Adapter

---

## Portfolio Objective Alignment - 2026-05-31

This issue is part of the Apple Silicon MLX REAP backend effort. The production/official REAP experimentation workflow remains the original PyTorch/CUDA implementation; this MLX work must not destabilize or replace that path.

The MLX backend goal is adapter-driven support for MoE weights, not a one-model port. Qwen3-specific work is bootstrap/reference-adapter work. The implementation should preserve REAP pruning semantics and observer-data schema compatibility while adapting to MLX constraints such as lazy execution, no PyTorch hooks, MLX-LM module layouts, and save/reload behavior.

For this issue, Mixtral is a second reference adapter demonstrating that the MLX design supports multiple routing semantics; it is not the end of architecture support.

**Comment by egesabanci** [comment]
Closing this as not planned; Mixtral adapter work is no longer in scope.
